src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/LSPSemanticTokensHighlightVisitor.java

```diff
@@ -53,22 +58,49 @@ public boolean suitableForFile(@NotNull PsiFile file) {
         return LanguageServersRegistry.getInstance().isFileSupported(file);
     }
 
+    private HighlightInfoHolder holder;
+    private LazyHighlightInfo[] lazyInfos;
+
     @Override
     public boolean analyze(@NotNull PsiFile file, boolean updateWholeFile, @NotNull HighlightInfoHolder holder, @NotNull Runnable action) {
         if (ProjectIndexingManager.canExecuteLSPFeature(file) != ExecuteLSPFeatureStatus.NOW) {
             return true;
         }
-        action.run();
-        // run unconditionally, because the LSP semanticTokens API sucks and is file-level only
-        highlightSemanticTokens(file, holder);
+        try {
+            if (SimpleLanguageUtils.isSupported(file.getLanguage())) {
+                highlightSemanticTokens(file, holder);
+                this.lazyInfos = null;
+                this.holder = null;
+            } else {
+                this.lazyInfos = highlightSemanticTokens(file, null);
+                this.holder = holder;
+            }
+            action.run();

```

Comment:
Is there any reason why action.run() is called at the end although before it was called at first?